The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10167800532/job/28120976943) [(plain)](https://github.com/rust-lang/rust/commit/120793aaa446b511710a3c85ef48c74582090aa2/checks/28120976943/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.752 Building wheels for collected packages: reuse
#13 2.753   Building wheel for reuse (pyproject.toml): started
#13 3.007   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.008   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.008   Stored in directory: /tmp/pip-ephem-wheel-cache-db151mjc/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.011 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.418 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.419 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.961 Collecting virtualenv
#13 3.961 Collecting virtualenv
#13 4.003   Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
#13 4.106      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 56.5 MB/s eta 0:00:00
#13 4.145 Collecting distlib<1,>=0.3.7
#13 4.151   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.158      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 91.9 MB/s eta 0:00:00
#13 4.190 Collecting filelock<4,>=3.12.2
#13 4.194   Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
#13 4.223 Collecting platformdirs<5,>=3.9.1
#13 4.227   Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
#13 4.309 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.498 Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
#13 DONE 4.6s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      198592 kB
DirectMap2M:     7141376 kB
DirectMap1G:    11534336 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/006c8df322e55c14d845e1fe317ca1445c2f8e6b/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-006c8df322e55c14d845e1fe317ca1445c2f8e6b-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
   Compiling tidy v0.1.0 (/checkout/src/tools/tidy)
    Finished `release` profile [optimized] target(s) in 30.21s
##[endgroup]
fmt check
##[error]Diff in /checkout/library/std/src/io/buffered/bufreader.rs at line 95:
         BufReader { inner, buf: Buffer::with_capacity(capacity) }
 
-    
-    
     /// Attempt to look ahead `n` bytes.
     ///
     /// `n` must be less than `capacity`.
##[error]Diff in /checkout/library/std/src/io/buffered/bufreader/buffer.rs at line 99:
     pub fn read_more(&mut self, mut reader: impl Read) -> io::Result<()> {
     pub fn read_more(&mut self, mut reader: impl Read) -> io::Result<()> {
         let mut buf = BorrowedBuf::from(&mut self.buf[self.pos..]);
-        let old_init = self.initialized-self.pos;
+        let old_init = self.initialized - self.pos;
         unsafe {
             buf.set_init(old_init);
         }
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_lint/src/map_unit_fn.rs" "/checkout/compiler/rustc_lint/src/expect.rs" "/checkout/compiler/rustc_lint/src/invalid_from_utf8.rs" "/checkout/compiler/rustc_lint/src/non_ascii_idents.rs" "/checkout/compiler/rustc_lint/src/nonstandard_style.rs" "/checkout/compiler/rustc_lint/src/reference_casting.rs" "/checkout/compiler/rustc_lint/src/context.rs" "/checkout/compiler/rustc_lint/src/deref_into_dyn_supertrait.rs" "/checkout/compiler/rustc_lint/src/early.rs" "/checkout/compiler/rustc_lint/src/noop_method_call.rs" "/checkout/compiler/rustc_lint/src/async_fn_in_trait.rs" "/checkout/compiler/rustc_lint/src/for_loops_over_fallibles.rs" "/checkout/compiler/rustc_lint/src/multiple_supertrait_upcastable.rs" "/checkout/compiler/rustc_lint/src/traits.rs" "/checkout/library/std/src/sync/mpsc/tests.rs" "/checkout/library/std/src/sync/mpsc/sync_tests.rs" "/checkout/library/std/src/sync/mpsc/mod.rs" "/checkout/library/std/src/sync/mpmc/utils.rs" "/checkout/library/std/src/sync/mpmc/select.rs" "/checkout/library/std/src/sync/mpmc/array.rs" "/checkout/library/std/src/sync/mpmc/waker.rs" "/checkout/library/std/src/sync/mpmc/counter.rs" "/checkout/library/std/src/sync/mpmc/context.rs" "/checkout/library/std/src/sync/mpmc/list.rs" "/checkout/library/std/src/sync/mpmc/mod.rs" "/checkout/library/std/src/sync/mpmc/error.rs" "/checkout/library/std/src/sync/mpmc/zero.rs" "/checkout/library/std/src/sync/rwlock.rs" "/checkout/library/std/src/sync/poison.rs" "/checkout/library/std/src/sync/mod.rs" "/checkout/library/std/src/sync/lazy_lock.rs" "/checkout/library/std/src/sync/once/tests.rs" "/checkout/library/std/src/sync/lazy_lock/tests.rs" "/checkout/library/std/src/sync/once.rs" "/checkout/library/std/src/pipe.rs" "/checkout/library/std/src/panicking.rs" "/checkout/library/std/src/io/buffered/bufwriter.rs" "/checkout/library/std/src/io/buffered/linewriter.rs" "/checkout/library/std/src/io/buffered/tests.rs" "/checkout/library/std/src/io/buffered/linewritershim.rs" "/checkout/library/std/src/io/buffered/bufreader.rs" "/checkout/library/std/src/io/buffered/bufreader/buffer.rs" "/checkout/library/std/src/io/buffered/mod.rs" "/checkout/library/std/src/io/copy/tests.rs" "/checkout/library/std/src/io/impls/tests.rs" "/checkout/compiler/rustc_data_structures/src/unord.rs" "/checkout/library/std/src/io/stdio.rs" "/checkout/compiler/rustc_data_structures/src/marker.rs" "/checkout/compiler/rustc_data_structures/src/captures.rs" "/checkout/compiler/rustc_data_structures/src/memmap.rs" "/checkout/compiler/rustc_data_structures/src/unhash.rs" "/checkout/compiler/rustc_data_structures/src/owned_slice.rs" "/checkout/library/std/src/io/stdio/tests.rs" "/checkout/library/std/src/io/prelude.rs" "/checkout/library/std/src/io/tests.rs" "/checkout/library/std/src/io/copy.rs" "/checkout/compiler/rustc_data_structures/src/intern/tests.rs" "/checkout/compiler/rustc_data_structures/src/snapshot_map/tests.rs" "/checkout/compiler/rustc_data_structures/src/snapshot_map/mod.rs" "/checkout/library/std/src/io/error/repr_bitpacked.rs" "/checkout/compiler/rustc_data_structures/src/small_c_str.rs" "/checkout/library/std/src/io/error/tests.rs" "/checkout/library/std/src/io/error/repr_unpacked.rs" "/checkout/compiler/rustc_lint/src/redundant_semicolon.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Tue Jul 30 18:35:10 UTC 2024
  network time: Tue, 30 Jul 2024 18:35:10 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/library/std/src/io/buffered/bufreader.rs
+++ b/library/std/src/io/buffered/bufreader.rs
@@ -94,6 +94,40 @@ impl<R: Read> BufReader<R> {
     pub fn with_capacity(capacity: usize, inner: R) -> BufReader<R> {
         BufReader { inner, buf: Buffer::with_capacity(capacity) }
     }
+
+    /// Attempt to look ahead `n` bytes.
+    ///
+    /// `n` must be less than `capacity`.
+    ///
+    /// ## Examples
+    ///
+    /// ```rust
+    /// #![feature(bufreader_peek)]
+    /// use std::io::{Read, BufReader};
+    ///
+    /// let mut bytes = &b"oh, hello"[..];
+    /// let mut rdr = BufReader::with_capacity(6, &mut bytes);
+    /// assert_eq!(rdr.peek(2).unwrap(), b"oh");
+    /// let mut buf = [0; 4];
+    /// rdr.read(&mut buf[..]).unwrap();
+    /// assert_eq!(&buf, b"oh, ");
+    /// assert_eq!(rdr.peek(2).unwrap(), b"he");
+    /// let mut s = String::new();
+    /// rdr.read_to_string(&mut s).unwrap();
+    /// assert_eq!(&s, "hello");
+    /// ```
+    #[unstable(feature = "bufreader_peek", issue = "128405")]
+    pub fn peek(&mut self, n: usize) -> io::Result<&[u8]> {
+        assert!(n <= self.capacity());
+        while n > self.buf.buffer().len() {
+            if self.buf.pos() > 0 {
+                self.buf.backshift();
+            }
+            self.buf.read_more(&mut self.inner)?;
+            debug_assert_eq!(self.buf.pos(), 0);
+        }
+        Ok(&self.buf.buffer()[..n])
+    }
 }
 
 impl<R: ?Sized> BufReader<R> {

--- a/library/std/src/io/buffered/bufreader/buffer.rs
+++ b/library/std/src/io/buffered/bufreader/buffer.rs
@@ -97,6 +97,27 @@ impl Buffer {
         self.pos = self.pos.saturating_sub(amt);
     }
 
+    /// Read more bytes into the buffer without discarding any of its contents
+    pub fn read_more(&mut self, mut reader: impl Read) -> io::Result<()> {
+        let mut buf = BorrowedBuf::from(&mut self.buf[self.pos..]);
+        let old_init = self.initialized - self.pos;
+        unsafe {
+            buf.set_init(old_init);
+        }
+        reader.read_buf(buf.unfilled())?;
+        self.filled += buf.len();
+        self.initialized += buf.init_len() - old_init;
+        Ok(())
+    }
+
+    /// Remove bytes that have already been read from the buffer.
+    pub fn backshift(&mut self) {
+        self.buf.copy_within(self.pos.., 0);
+        self.initialized -= self.pos;
+        self.filled -= self.pos;
+        self.pos = 0;
+    }
+
     #[inline]
     pub fn fill_buf(&mut self, mut reader: impl Read) -> io::Result<&[u8]> {
         // If we've reached the end of our internal buffer then we need to fetch

